ordered_product/models.py

```diff
@@ -0,0 +1,63 @@
+from django.db import models
+from django.core.validators import MinValueValidator
+from django.core.exceptions import ValidationError
+from product.models import Product
+from supplier.models import Supplier
+from client.models import Client
+from delivery_location.models import DeliveryLocation
+
+
+class SupplierProduct(models.Model):
+    qr_code = models.ForeignKey(Product, on_delete=models.CASCADE)
+    user_name = models.ForeignKey(Supplier, on_delete=models.CASCADE)
+    price = models.IntegerField(validators=[MinValueValidator(1)])
+    quantity = models.IntegerField(validators=[MinValueValidator(1)])
+
+
+class OrderedProduct(models.Model):
+    delivery_location_id = models.ForeignKey(DeliveryLocation, on_delete=models.CASCADE)
+    user_name = models.ForeignKey(Client, on_delete=models.CASCADE)
+    supplier_product_id = models.ForeignKey(SupplierProduct, on_delete=models.CASCADE)
+    quantity = models.IntegerField(default=1, validators=[MinValueValidator(1)])
+
+    @staticmethod
+    def filter_user_name(user_name):
+        return OrderedProduct.objects.filter(user_name=user_name) if user_name else []
+
+    @staticmethod
+    def filter_delivery_location_id(dl_id):
+        return OrderedProduct.objects.filter(delivery_location_id=dl_id) if dl_id else []
+
+    @staticmethod
+    def filter_supplier_product_id(sp_id):
+        return OrderedProduct.objects.filter(supplier_product_id=sp_id) if sp_id else []
+
+    def save_ordered_product(self):
+        self.save()
+        try:
+            OrderedProduct.full_clean(self)
+        except ValidationError as e:
+            self.delete()
+            raise e
+        return self
+
+    def delete_ordered_product(self):
+        OrderedProduct.objects.get(id=self.id).delete()
+
+    def decrease_quantity(self, qn):
+        self.quantity -= qn
+        self.save()
+        if (self.quantity <= 0):
```

Comment:
the `save` line is potentially putting the DB in an invalid state until you delete it. it means a race condition in which others can access invalid data between the time you save it and the time you delete it.

Honestly, any way you turn this around, there's going to be some sort of race condition here unless you make this method atomic (e.g. you decrement a value different than the one you read, etc.), and I don't want you to spend time on that. so, for now, I think you can leave it as-is, but if someone asks you about it one day, just know there's an issue here :)

Reply:
(All in all, nothing to do here for now, just FYI for whoever reads)